pkg/controller/deployment/deployment_controller.go

```diff
@@ -630,9 +630,73 @@ func (dc *DeploymentController) reconcileOldRCs(allRCs []*api.ReplicationControl
 		// Cant scale down further
 		return false, nil
 	}
+
+	// Check the expectations of deployment before reconciling
+	dKey, err := controller.KeyFunc(&deployment)
+	if err != nil {
+		return false, fmt.Errorf("Couldn't get key for deployment %#v: %v", deployment, err)
+	}
+	if expectationsCheck && !dc.podExpectations.SatisfiedExpectations(dKey) {
+		glog.V(4).Infof("Pod expectations not met yet before reconciling old RCs\n")
+		return false, nil
+	}
+
+	// Clean up all unhealthy replicas, otherwise unhealthy replicas will block deployment
+	// and cause timeout. See https://github.com/kubernetes/kubernetes/issues/16737
+	cleanupCount, err := dc.cleanupUnhealthyReplicas(oldRCs, deployment)
+	if err != nil {
+		return false, nil
+	}
+
+	// Scale down old rcs, need check maxUnavailable to ensure we can scale down
+	scaledDown, err := dc.scaleDownOldRCsForRollingUpdate(allRCs, oldRCs, deployment)
+	if err != nil {
+		return false, nil
+	}
+
+	totalScaledDown := cleanupCount + scaledDown
+	if expectationsCheck {
+		dc.podExpectations.ExpectDeletions(dKey, totalScaledDown)
+	}
+
+	return totalScaledDown > 0, nil
+}
+
+// cleanupUnhealthyReplicas will scale down old rcs with unhealthy replicas, so that all unhealthy replicas will be deleted.
+func (dc *DeploymentController) cleanupUnhealthyReplicas(oldRCs []*api.ReplicationController, deployment extensions.Deployment) (int, error) {
```

Comment:
BTW, we should document what I wrote above, in code comments and also in the design doc.